storage/filesystem.go

```diff
@@ -223,14 +252,42 @@ func (f *Filesystem) tempfsPath() string {
 	return filepath.Join(f.baseDir, "inmemfs")
 }
 
-func makePayload(in map[string][]byte) map[string]util.FileProjection {
+func makePayload(in map[string][]byte, fsGroup *int64) map[string]util.FileProjection {
 	out := make(map[string]util.FileProjection, len(in))
 	for name, data := range in {
 		out[name] = util.FileProjection{
-			Data: data,
-			// read-only for user + group (TODO: set fsUser/fsGroup)
-			Mode: readOnlyUserAndGroupFileMode,
+			Data:    data,
+			FsGroup: fsGroup,
+			Mode:    readOnlyUserAndGroupFileMode,
 		}
 	}
 	return out
 }
+
+// fsGroupForMetadata returns the gid that ownership of the volume data
+// directory should be changed to. Returns nil if ownership should not be
+// changed.
+func (f *Filesystem) fsGroupForMetadata(meta metadata.Metadata) (*int64, error) {
+	// FixedFSGroup takes precedence over attribute key.
+	if f.FixedFSGroup != nil {
+		return f.FixedFSGroup, nil
+	}
+
+	// If the FSGroupVolumeAttributeKey is not defined, no ownership can change.
+	if f.FSGroupVolumeAttributeKey == nil {
+		return nil, nil
+	}
+
+	fsGroupStr, ok := meta.VolumeContext[*f.FSGroupVolumeAttributeKey]
+	if !ok {
+		// If the attribute has not been set, return no ownership change.
+		return nil, nil
+	}
+
+	fsGroup, err := strconv.ParseInt(fsGroupStr, 10, 64)
```

Comment:
Should we also validate that the given gid is also a valid group ID? i.e. >0, perhaps less than some value too?

Reply:
Yep, I agree. Will add these check.